src/cfnlint/rules/resources/stepfunctions/StateMachine.py

```diff
@@ -0,0 +1,155 @@
+            'Pass': [],
+            'Task': ['Resource'],
+            'Choices': ['Choices'],
+            'Wait': [],
+            'Succeed': [],
+            'Fail': [],
+            'Parallel': ['Branches']
+        }
+
+        for req_key in common_state_required_keys:
+            if req_key not in def_json:
+                message = 'State Machine Definition required key (%s) for State (%s) is missing' % (req_key, state_name)
+                matches.append(RuleMatch(path, message))
+                return matches
+
+        state_type = def_json.get('Type')
+
+        for state_key, _ in def_json.items():
+            if state_key not in common_state_keys + state_key_types.get(state_type):
+                message = 'State Machine Definition key (%s) for State (%s) of Type (%s) is not valid' % (state_key, state_name, state_type)
+                matches.append(RuleMatch(path, message))
+        for req_key in common_state_required_keys + state_required_types.get(state_type):
+            if req_key not in def_json:
+                message = 'State Machine Definition required key (%s) for State (%s) of Type (%s) is missing' % (req_key, state_name, state_type)
+                matches.append(RuleMatch(path, message))
+                return matches
+
+        return matches
+
+    def _check_definition_json(self, def_json, path):
+        """Check JSON Definition"""
+        matches = list()
+
+        top_level_keys = [
+            'Comment',
+            'StartAt',
+            'TimeoutSeconds',
+            'Version',
+            'States'
+        ]
+        top_level_required_keys = [
+            'StartAt',
+            'States'
+        ]
+        for top_key, _ in def_json.items():
+            if top_key not in top_level_keys:
+                message = 'State Machine Definition key (%s) is not valid' % top_key
+                matches.append(RuleMatch(path, message))
+
+        for req_key in top_level_required_keys:
+            if req_key not in def_json:
+                message = 'State Machine Definition required key (%s) is missing' % req_key
+                matches.append(RuleMatch(path, message))
+
+        for state_name, state_value in def_json.get('States', {}).items():
+            matches.extend(self._check_state_json(state_value, state_name, path))
+        return matches
+
+    def check_value(self, value, path):
+        """Check Definition Value"""
+        matches = list()
+        try:
+            def_json = json.loads(value)
+        # pylint: disable=W0703
+        except Exception as err:
+            message = 'State Machine Definition needs to be formatted as JSON. Error %s' % err
+            matches.append(RuleMatch(path, message))
+            return matches
+
+        matches.extend(self._check_definition_json(def_json, path))
+        return matches
+
+    def check_sub(self, value, path):
+        """Check Sub Object"""
+        matches = list()
+        if isinstance(value, list):
```

Comment:
Trying to understand under what circumstances this might be a list and why we only need the first value if it is?

Reply:
My solutions to this would either be:
- not check the state definitions when Fn::Sub is used
- create a way to determine if someone is doing the more complex method and not check it. 
  - JSON blobs addended to the state definition (this may be harder to detect as it may be based on the parse-ability)  This should be really uncommon and in this scenario the rule should be ignored.
  - Using Sub variables for JSON key names.  Easier to detect as it should still be parseable and we would check the key values for things that looked like variables.  

I'm leaning the later as I would expect Subs would be very common for this.

Reply:
After talking, it sounds like there are edge cases but this should solve for 80/20 -- and the 20 can just disable the rule and file an issue.  Rest of the rule looks good.